packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp

```diff
@@ -70,17 +72,181 @@ class Distributor;
 //
 namespace Details {
 
-template<class LocalSparseMatrixType>
+
+/// \brief Compute number of packets per row of the local matrix
+///
+/// \tparam GO Global ordinal type
+/// \tparam ViewType Type of the Kokkos::View specialization
+///   used to store the number of packets; the output array of this function.
+/// \tparam ExportLIDsViewType Type of the Kokkos::View specialization
+///   used to store the LIDs of local rows; the input array of this function.
+/// \tparam LocalMatrixType Type of the KokkosSparse::CrsMatrix local matrix
+///
+template<class GO, class ViewType, class ExportLIDsViewType, class LocalMatrixType>
+struct NumPacketsFunctor {
+
+  typedef typename ViewType::value_type value_type;
+  typedef typename LocalMatrixType::row_map_type row_map_type;
+  typedef typename LocalMatrixType::value_type IST;
+  typedef typename LocalMatrixType::ordinal_type LO;
+
+  ViewType num_packets;
+  ExportLIDsViewType export_lids;
+  row_map_type row_map;
+
+  NumPacketsFunctor(ViewType num_packets_, ExportLIDsViewType export_lids_,
+      LocalMatrixType local_matrix):
+    num_packets(num_packets_), export_lids(export_lids_),
+    row_map(local_matrix.graph.row_map) {};
+
+  KOKKOS_INLINE_FUNCTION
+    void operator() (size_t i) const {
+      LO export_lid = static_cast<LO>(export_lids[i]);
+      const size_t count = static_cast<size_t>(row_map[export_lid+1]
+                                             - row_map[export_lid]);
+      const value_type num_bytes = sizeof(LO) + count * (sizeof(IST) + sizeof(GO));
+      num_packets[i] = (count == 0) ? 0 : num_bytes;
+    }
+};
+
+/// \brief Reduction result for PackCrsMatrixFunctor below.
+///
+/// The reduction result finds the offset and number of bytes associated with
+/// the first out of bounds error or packing error occurs.
+struct PackCrsMatrixError {
+
+  size_t bad_index; // only valid if outOfBounds == true.
+  size_t bad_offset;   // only valid if outOfBounds == true.
+  size_t bad_num_bytes; // only valid if outOfBounds == true.
+  bool out_of_bounds_error;
+  bool packing_error;
+
+  PackCrsMatrixError():
+    bad_index(std::numeric_limits<size_t>::max()),
+    bad_offset(0),
+    bad_num_bytes(0),
+    out_of_bounds_error(false),
+    packing_error(false) {}
+
+  KOKKOS_INLINE_FUNCTION bool success() const
+  {
+    return ((!out_of_bounds_error) && (!packing_error));
+  }
+
+  KOKKOS_INLINE_FUNCTION std::string summary() const
```

Comment:
You don't need to call this in a thread-parallel way, and it shouldn't be (because it does I/O), so don't mark it `KOKKOS_INLINE_FUNCTION`.